File: Automation_App_Frontend/lib/features/mandanten/domain/entities/ordnernamen_menge.dart
import 'package:equatable/equatable.dart';

/// Eine Menge von Ordnernamen, die **ohne Rücksicht auf Groß- und
/// Kleinschreibung** vergleicht.
///
/// Ordnernamen kommen aus dem Windows-Dateisystem, und dort sind
/// „VUnfallursache Mark" und „Vunfallursache Mark" derselbe Ordner — zwei
/// solche nebeneinander gibt es gar nicht. Genau verglichen hinge es an der
/// Schreibweise der Quelle, ob ein zugeordneter Ordner im Zuordnungsstapel
/// stehen bleibt und ob ein Vermerk seinen Ordner wiederfindet; im
/// schlimmsten Fall wäre ein Ordner zugeordnet **und** „ohne Mandantenbezug".
/// Dieselbe Regel gilt im Backend: `OrdnerStatusRegister` vergleicht
/// `OrdinalIgnoreCase`, die Spalte trägt die Kollation `NOCASE`.
class OrdnernamenMenge extends Equatable {
  final Set<String> _kleingeschrieben;

  OrdnernamenMenge(Iterable<String> ordnernamen)
    : _kleingeschrieben = {
        for (final name in ordnernamen) name.trim().toLowerCase(),
      };

  /// Ob [ordnername] in der Menge steht — gleich, wie er geschrieben ist.
  bool enthaelt(String ordnername) =>
      _kleingeschrieben.contains(ordnername.trim().toLowerCase());

  /// [ordnernamen] ohne die, die in der Menge stehen — in ihrer Reihenfolge.
  /// Der eine Weg, Namen aus einer Liste zu nehmen: Von Hand ausgeschrieben
  /// stand dieselbe Schleife an vier Stellen, und die nächste hätte beim
  /// Vergleich vielleicht nicht getrimmt.
  List<String> ohne(Iterable<String> ordnernamen) =>
      ohneEintraege(ordnernamen, (name) => name);

  /// Wie [ohne], für Einträge, die einen Ordnernamen tragen —
  /// [ordnernameVon] liest ihn.
  List<T> ohneEintraege<T>(
    Iterable<T> eintraege,
    String Function(T eintrag) ordnernameVon,
  ) => [
    for (final eintrag in eintraege)
      if (!enthaelt(ordnernameVon(eintrag))) eintrag,
  ];

  int get length => _kleingeschrieben.length;

  bool get isEmpty => _kleingeschrieben.isEmpty;

  bool get isNotEmpty => _kleingeschrieben.isNotEmpty;

  @override
  List<Object?> get props => [_kleingeschrieben];
}

File: AutomationService/AutomationService/Features/Mandanten/Domain/Services/OrdnernamenMenge.cs
namespace AutomationService.Features.Mandanten.Domain.Services;

/// <summary>
/// Eine Menge von Akten-Ordnernamen, die <b>getrimmt und ohne Rücksicht auf
/// Groß-/Kleinschreibung</b> vergleicht — dieselbe Regel wie
/// <c>OrdnernamenMenge</c> im Frontend. Der Name kommt aus dem
/// Windows-Dateisystem, und dort meinen „VUnfallursache Mark" und
/// „Vunfallursache Mark " denselben Ordner.
///
/// Getrimmt wird auf <b>beiden</b> Seiten: Kürzte nur die geprüfte Seite,
/// gälte ein gespeichertes „ Foo" im Backend als frei und im Frontend als
/// vergeben — die beiden Schichten wären sich uneins, wem ein Ordner gehört.
/// </summary>
public sealed class OrdnernamenMenge(IEnumerable<string> ordnernamen)
{
    readonly HashSet<string> _namen = new(
        ordnernamen.Select(Schluessel), StringComparer.OrdinalIgnoreCase);

    /// <summary>Ob <paramref name="ordnername"/> in der Menge steht — gleich, wie er geschrieben ist.</summary>
    public bool Enthaelt(string ordnername) => _namen.Contains(Schluessel(ordnername));

    /// <summary>
    /// Die Vergleichsform eines Ordnernamens. Wer Namen als Schlüssel eines
    /// Wörterbuchs mit <see cref="StringComparer.OrdinalIgnoreCase"/> ablegt,
    /// nimmt sie für Ablage <b>und</b> Nachschlagen.
    /// </summary>
    public static string Schluessel(string ordnername) => ordnername.Trim();
}
